crates/bevy_ecs/src/storage/table/mod.rs

```diff
@@ -716,32 +501,83 @@ impl Table {
 
     /// Reserves `additional` elements worth of capacity within the table.
     pub(crate) fn reserve(&mut self, additional: usize) {
-        if self.entities.capacity() - self.entities.len() < additional {
+        if self.capacity() - self.len() < additional {
+            let column_cap = self.capacity();
             self.entities.reserve(additional);
 
             // use entities vector capacity as driving capacity for all related allocations
             let new_capacity = self.entities.capacity();
 
-            for column in self.columns.values_mut() {
-                column.reserve_exact(new_capacity - column.len());
+            if column_cap == 0 {
+                // SAFETY: 0 < `column_cap` <= `new_capacity`
+                unsafe { self.alloc_columns(NonZeroUsize::new_unchecked(new_capacity)) };
+            } else {
+                // SAFETY:
+                // - `column_cap` is indeed the columns' capacity
+                // - 0 < `additional` <= `self.len() + additional` <= `new_capacity`
+                unsafe {
+                    self.realloc_columns(
+                        NonZeroUsize::new_unchecked(column_cap),
+                        NonZeroUsize::new_unchecked(new_capacity),
+                    );
+                };
             }
         }
     }
 
+    /// Allocate memory for the columns in the [`Table`]
+    ///
+    /// The current capacity of the columns should be 0, if it's not 0, then the previous data will be overwritten and leaked.
+    fn alloc_columns(&mut self, new_capacity: NonZeroUsize) {
+        self.poisoned = true;
```

Comment:
If we panic on allocation, I think we should just abort. This should be doable by panicking again while we're unwinding. Namely, panicking inside of a Drop impl. An unsynchronized column will cause out-of-bounds accesses, so this poisoning is unsound unless we check for it on every access, which is probably not what we want to do.

Reply:
Makes sense. But I don't really understand how to implement it, could you expand further on it? based on what condition would we trigger a panic in the `Drop` impl? Or maybe just replace `handle_alloc_error` with `std::process::abort`?

for context, I implemented this "poison" method after this comment: https://github.com/bevyengine/bevy/pull/12929#discussion_r1564470891

Reply:
```rust
struct AbortOnPanic;

impl Drop for AbortOnPanic {
  fn drop(&mut self)  {
       // Panicking while panicking will force an abort.
       panic!("Aborting due to allocator error");
   }
}

let _guard = AbortOnPanic;

// do things you need

std::mem::forget(_guard);
```